src/core/reference/include/openvino/reference/paged_attention.hpp

```diff
@@ -0,0 +1,330 @@
+            trig_index = rotation_deltas[rotated_index];
+        else if (rotation_deltas_shape[1] == static_cast<size_t>(block_size))
+            trig_index = rotation_deltas[rotated_index * block_size + token_offset];
+    }
+    return trig_index;
+}
+
+/*
+ * Check if rotation should be applied.
+ * Returns true if rotated_block_indices is not null, the number of rotated blocks is > 0,
+ * and the block_id is found in the rotated list.
+ */
+bool should_rotate(int32_t block_id,
+                   const int32_t* rotated_block_indices,
+                   const ov::Shape& rotated_block_indices_shape,
+                   int32_t& rotated_index) {
+    int32_t num_rotated_blocks =
+        !rotated_block_indices_shape.empty() ? static_cast<int32_t>(rotated_block_indices_shape[0]) : 0;
+    if (rotated_block_indices && num_rotated_blocks > 0) {
+        for (int32_t i = 0; i < num_rotated_blocks; i++) {
+            if (rotated_block_indices[i] == block_id) {
+                rotated_index = i;
+                return true;
+            }
+        }
+    }
+    return false;
+}
+}  // namespace paged_attention_utils
+
+template <typename T>
+void paged_attention(
+    T* out,                               // output: attention result
+    T* score,                             // output: concatenated raw scores
+    const T* query,                       // shape: [batch_tokens, num_heads * head_size]
+    const T* key,                         // shape: [batch_tokens, num_kv_heads * head_size]
+    const T* value,                       // shape: [batch_tokens, num_kv_heads * head_size]
+    const T* key_cache,                   // shape: [num_blocks, num_kv_heads, block_size, head_size]
+    const T* value_cache,                 // shape: [num_blocks, num_kv_heads, block_size, head_size]
+    const Shape& q_shape,                 // e.g. {batch_tokens, num_heads * head_size}
+    const Shape& kv_shape,                // e.g. {batch_tokens, num_kv_heads * head_size}
+    const Shape& kv_cache_shape,          // e.g. {num_blocks, num_kv_heads, block_size, head_size}
+    const int32_t* past_lens,             // [batch_seq]: past tokens per sequence
+    const Shape& past_lens_shape,         // e.g. {batch_seq}
+    const int32_t* subsequence_begins,    // [batch_seq + 1]: start indices of new tokens per sequence
+    const int32_t* block_indices,         // [num_blocks]: block table for each sequence
+    const int32_t* block_indices_begins,  // [batch_seq + 1]: indices into block_indices per sequence
+    const T* scale_ptr,                   // (Optional) attention scale factor (can be nullptr; default = 1)
+    const int32_t* sliding_window_ptr,    // (Optional) sliding window parameter (can be nullptr; default = 0)
+    const T* alibi_slopes,  // (Optional) [num_kv_heads]: per-head bias slopes (can be nullptr; default = 0)
+    const int32_t* max_context_len_ptr,  // max context length (for score output indexing)
```

Comment:
Consider some o f these parameters to some kind of trivial struct to reduce number of parameters.

Reply:
Done